### ISSUE TITLE
Document thread-safety in Godoc

### ISSUE THREAD

**Issue by cretz** [open-policy-agent/opa#2129]
Took me a while to find https://github.com/open-policy-agent/opa/issues/1602#issuecomment-516736394 when researching the library. By default when not mentioned in Godoc, people assume that it is not thread/goroutine-safe. I suggest at the top of the `rego` package or on important objects/functions, just document that the calls can be made safely across goroutines.

(I'd open a PR to do this myself, but I am not comfortable enough with making the general assertion of thread-safety myself)

**Comment by ashutosh-narkar** [comment]
FWIW the [Integrating with the Go API](https://www.openpolicyagent.org/docs/latest/integration/#integrating-with-the-go-api) section of the docs has a note about this. 

> Prepared queries are safe to share across multiple Go routines.

**Comment by skillcoder** [comment]
Yep I saw it  `Prepared queries are safe to share across multiple Go routines.`
But my test today catch this:
```
==================
WARNING: DATA RACE
Read at 0x00c005dc0d48 by goroutine 118:
  github.com/open-policy-agent/opa/ast.(*object).Hash()
      .../vendor/github.com/open-policy-agent/opa/ast/term.go:1844 +0x64
  github.com/open-policy-agent/opa/ast.(*Term).Hash()
     .../vendor/github.com/open-policy-agent/opa/ast/term.go:395 +0x71
  github.com/open-policy-agent/opa/ast.(*set).get()
      .../vendor/github.com/open-policy-agent/opa/ast/term.go:1611 +0x4f
...
Previous write at 0x00c005dc0d48 by goroutine 121:
  github.com/open-policy-agent/opa/ast.(*object).Hash()
      .../vendor/github.com/open-policy-agent/opa/ast/term.go:1847 +0x108
  github.com/open-policy-agent/opa/ast.(*Term).Hash()
      .../vendor/github.com/open-policy-agent/opa/ast/term.go:395 +0x71
  github.com/open-policy-agent/opa/ast.(*set).get()
      .../vendor/github.com/open-policy-agent/opa/ast/term.go:1611 +0x4f
...
```
`github.com/open-policy-agent/opa v0.29.4`

I run it according doc:
Prepared once:
```
	preparedQuery, err := rego.New(
		rego.Query("data.rbac.allow"),
		rego.Module("rule.rego", rule),
	).PrepareForEval(ctx)
```
And call in goroutines:
```
res, err := query.Eval(ctx, rego.EvalInput(toInput(req)))
```

**Comment by ashutosh-narkar** [comment]
Can you share your code that integrates with OPA ? [Here's](https://github.com/open-policy-agent/opa-envoy-plugin/blob/main/envoyauth/evaluation.go#L31) an example of this is used.

**Comment by skillcoder** [comment]
Still relevant, even in latest v0.37.2
@ashutosh-narkar Please, try to use this test to reproduce it
https://github.com/skillcoder/opa-race

**Comment by ashutosh-narkar** [comment]
https://github.com/open-policy-agent/opa/issues/4345 addresses the same. We can track the work in that issue.

**Comment by anderseknert** [comment]
Regal uses the `rego` API with high parallelism, and while we found some minor issues initially, we've had a very smooth experience for the past year or two. A few comment have also been added to functions where extra care needs to be taken with regards to thread safety. I think it's OK to close this as resolved now.
